compiler/ilgen/VirtualMachineOperandArray.cpp

```diff
@@ -0,0 +1,181 @@
+   : VirtualMachineState(),
+   _mb(mb),
+   _numberOfElements(numOfElements),
+   _elementType(elementType),
+   _arrayBaseRegister(arrayBaseRegister)
+   {
+   int32_t numBytes = _numberOfElements * sizeof(TR::IlValue *);
+   _values = (TR::IlValue **) TR::comp()->trMemory()->allocateHeapMemory(numBytes);
+   memset(_values, 0, numBytes);
+
+   // store current operand stack pointer base address so we can use it whenever we need
+   // to recreate the stack as the interpreter would have
+   mb->Store("OperandArray_base", arrayBaseRegister->Load(mb));
+   }
+
+VirtualMachineOperandArray::VirtualMachineOperandArray(OMR::VirtualMachineOperandArray *other)
+   : VirtualMachineState(),
+   _mb(other->_mb),
+   _numberOfElements(other->_numberOfElements),
+   _elementType(other->_elementType),
+   _arrayBaseRegister(other->_arrayBaseRegister)
+   {
+   int32_t numBytes = _numberOfElements * sizeof(TR::IlValue *);
+   _values = (TR::IlValue **) TR::comp()->trMemory()->allocateHeapMemory(numBytes);
+   memcpy(_values, other->_values, numBytes);
+   }
+
+
+// commits the simulated operand array of values to the virtual machine state
+// the given builder object is where the operations to commit the state will be inserted
+// into the array which is assumed to be managed independently, most likely
+void
+VirtualMachineOperandArray::Commit(TR::IlBuilder *b)
+   {
+   TR::IlType *Element = _elementType;
+   TR::IlType *pElement = _mb->typeDictionary()->PointerTo(Element);
+
+   TR::IlValue *arrayBase = b->Load("OperandArray_base");
+
+   for (int32_t i = 0;i < _numberOfElements;i++)
+      {
+      TR::IlValue *element = _values[i];
+      if (element != NULL)
+         {
+         b->StoreAt(
+         b->   IndexAt(pElement,
+                  arrayBase,
+         b->      ConstInt32(i)),
+               element);
+         }
+      }
+   }
+
+void
+VirtualMachineOperandArray::Reload(TR::IlBuilder* b)
+   {
+   TR::IlType* Element = _elementType;
+   TR::IlType* pElement = _mb->typeDictionary()->PointerTo(Element);
+   // reload the elements back into the simulated operand array
+   TR::IlValue* array = b->Load("OperandArray_base");
+   for (int32_t i = 0; i < _numberOfElements; i++)
+      {
+      _values[i] = b->LoadAt(pElement,
+                   b->   IndexAt(pElement,
+                            array,
+                   b->      ConstInt32(i)));
+       }
+   }
+
+void
+VirtualMachineOperandArray::MergeInto(OMR::VirtualMachineState *o, TR::IlBuilder *b)
+   {
+   VirtualMachineOperandArray *other = (VirtualMachineOperandArray *)o;
+   TR_ASSERT(_numberOfElements == other->_numberOfElements, "array are not same size");
+   for (int32_t i=0;i < _numberOfElements;i++)
+      {
+      if (NULL == other->_values[i])
+         {
+         TR_ASSERT(_values[i] == NULL, "if an element in the dest array at index is NULL it has to be NULL in the src array");

```

Comment:
This assert establishes a "static" rule that all operands uses must be reached by an earlier definition on every incoming path. I think this is the only reasonable choice, but I wanted to call out that this assertion codifies it (but projects that leverage VirtualMachineOperandArray, anyway :) ).